RECIPE-4 Edit Social Media Profile

PatsiChang opened the pull request:
Link: https://smart-home-cloud.atlassian.net/browse/RECIPE-4
Link: https://smart-home-cloud.atlassian.net/browse/RECIPE-32
Changes: Added Backend for Smart Home Social Media 
Changes: Fixed token for Recipe Flow

PatsiChang committed:
Social Media restful Api and bean

Added Social Media bean and restful api

ChakShingLo commented on src/main/java/com/patsi/bean/SocialMediaUser.java:
```diff
@@ -1,11 +1,10 @@
-import jakarta.persistence.Entity;
-import jakarta.persistence.Id;
+import jakarta.persistence.*;
 import lombok.*;
```
best practice, avoid using * in import

PatsiChang replied:
oki

PatsiChang replied:
oki

ChakShingLo replied:
think you haven’t  change it

ChakShingLo commented on src/main/java/com/patsi/bean/SocialMediaUser.java:
```diff
@@ -14,23 +13,23 @@
-    //Stores the userName of the followed accounts
-    private List<String> followedUsers;
+    private int followersCount;
+    private int followingCount;
```
What is the difference between followersCount and followingCount?

PatsiChang replied:
one is count of followers one is count of how many people the user follows

ChakShingLo commented on src/main/java/com/patsi/bean/UserFollows.java:
```diff
@@ -0,0 +1,36 @@
+@Entity
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
```
SUggestion: Try to use builder instead of ArgsConstructor?

PatsiChang replied:
oki

ChakShingLo commented on src/main/java/com/patsi/controller/SocialMediaController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @GetMapping
+    public SocialMediaUser getUserById(@RequestParam SocialMediaUser user) {
```
Why you need a all value of the SocialMediaUser in a get request?.
Think user id should be sufficient?

PatsiChang replied:
The explore page might need to get All Users?

ChakShingLo replied:
I mean user id should be sufficient enough for you to query the user. You don't need all data of the SocialMediaUser.

ChakShingLo commented on src/main/java/com/patsi/service/SocialMediaService.java:
```diff
@@ -0,0 +1,42 @@
+
+    //Delete GroceryItem
+    public boolean deactivateAccount(SocialMediaUser user) {
+        user.setAccountStatus(false);
```
Should we use enum for account status? I can imagine there will be a lot more status like suspended, active, closed, inactive, etc... in the future.

PatsiChang replied:
ok

ChakShingLo reviewed (commented):
Looks good to me. Just some question and suggestion

PatsiChang committed:
Fix Recipe 4 PR

Added enum type for Account Status, Added builder

PatsiChang committed:
Fix Account Status and Imports

Removed imports with * and fixed service and controller

ChakShingLo reviewed (approved):
LGTM

PatsiChang committed:
Added Social Media Flow

Add Social media flow, add token flow to recipe. (Todo: Need to fix all recipe flows for tokens)

PatsiChang committed:
Fix Social Media controller and service

Changed the directory of recipeIcon img and added directory for social media profile picture

ChakShingLo commented on src/main/java/com/patsi/bean/SocialMediaUser.java:
```diff
@@ -7,31 +7,35 @@
     private String biography;
-    private int followersCount;
-    private int followingCount;
+    private int followersCount = 0;
```
U may need @Builder.Default for default value

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/controller/RecipeController.java:
```diff
@@ -53,10 +59,17 @@ public void updateRecipeIcon(@RequestParam("recipeID") String recipeID,
-        log.info("Inside Controller Get Recipe");
-        return recipeService.getRecipe();
+    @PostMapping("getRecipeByToken")
+    public List<Recipe> getRecipe(@RequestBody String token) {
```
think better mark it TODO as we are going to include the token in HTTP header. Then a interceptor will include the token into something like RequestContext / HttpSessionContext

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/controller/SocialMediaUserController.java:
```diff
@@ -0,0 +1,92 @@
+    @Autowired
+    private SocialMediaService socialMediaService;
+
+    public String getUidFromToken ( String token ) {
```
Think this method should be extracted to a separated service. Let say UserProfile Service?

ChakShingLo replied:
Also, remove those extra spacing, should be:
`public String getUidFromToken(String token) {`

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/interceptors/LoggingInterceptor.java:
```diff
@@ -18,6 +18,7 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
         return true;
     }
 
+
```
remove this extra blank line?

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/service/RecipeService.java:
```diff
@@ -38,60 +39,68 @@ public UUID registerRecipe(Recipe recipe) {
-                outputStream.write(recipeIcon);
-            }
+    public void updateRecipeIcon(UUID id, byte[] recipeIcon) throws IOException {
+        File f = FileHelper.newFile(IMAGE_PATH + id + ".jpg");
```
think you can combine the whole writing file logic into one single method in FileHelper, which accept a file path + name and the byte[].

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/service/RecipeService.java:
```diff
@@ -38,60 +39,68 @@ public UUID registerRecipe(Recipe recipe) {
         }
-        int randomNum = (int)(Math.random()*returnList.size());
-        if (randomNum > 0){
+        int randomNum = (int) (Math.random() * returnList.size());
```
1. better create a RandomUtil, then a method wrapping Math.random()
2. create some util method which generate a random int within the given min/max

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/service/RecipeService.java:
```diff
@@ -38,60 +39,68 @@ public UUID registerRecipe(Recipe recipe) {
-            if (tmpRecipeList.size() > 0){
-                int randomNumWholeList = (int)(Math.random()*tmpRecipeList.size());
+            if (tmpRecipeList.size() > 0) {
+                int randomNumWholeList = (int) (Math.random() * tmpRecipeList.size());
```
same as the previous comment.
1. better create a RandomUtil, then a method wrapping Math.random()
2. create some util method which generate a random int within the given min/max

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/service/SocialMediaService.java:
```diff
@@ -18,26 +27,77 @@ public class SocialMediaService {
+        log.info("In Service: changeBannerPicture");
+        SocialMediaUser user = socialMediaRepository.findByUserName(userName).orElse(null);
+        if(user != null){
+            File f = FileHelper.newFile(IMAGE_PATH_BannerPicture + user.getUserName() +".jpg");
```
same here for extracting to file util.
`think you can combine the whole writing file logic into one single method in FileHelper, which accept a file path + name and the byte[].`

ChakShingLo reviewed (commented):
some comments

PatsiChang committed:
Added util and fixed codes

Fixed changes in PL recipe-4

ChakShingLo commented on src/main/java/com/patsi/utils/FileHelper.java:
```diff
@@ -5,14 +5,16 @@
-            return new FileOutputStream(f);
-        } catch (Exception e){
-            return null;
+    public static void newFile(String pathname, String profilePictureID, byte[] image)
```
if this is a method for new file only, it better to check that file does not exist.
throw exception if the file already exist. 
P.s. you can throw RuntimeException if you don't wanna force catching it

PatsiChang replied:
fixed

ChakShingLo commented on src/main/java/com/patsi/utils/GenerateIntHelper.java:
```diff
@@ -0,0 +1,9 @@
+
+public class GenerateIntHelper {
+    public static double generateRandomInt(int min, int max) {
+        return Math.floor(Math.random() * (max - min + 1) + min);
```
wrap the Math.random() with a separated method

PatsiChang replied:
fixed

PatsiChang committed:
Fixed utils

Fixed Math random and check if file exist

PatsiChang committed:
Changed method name for utils

Changed names for math generator method
